### PR TITLE
Improvements to support property tests on Traces with simple Tx with DRep related Certs

### LINKED ISSUE #3610: Systematic simple, but long traces
We need long, but simple traces to be able test things at an epoch boundary and other functionality that depends on long running logic (pulsers, incremental computation, etc.). 

DRep distribution is the first test candidate for this.

### PR DESCRIPTION
**Pull request by TimSheard**
Added features for chains of simple transactions.
Added Univ for DRep and DRepState,
Better constraints for VState, Added constraints for EpochState and NewEpochState.
New vars for Conway PParams, and needed (Rep era t). 
Added missing Conway PParams fields, added abstractPParams, and pcParams. 
Added module DrepCertTx which generates simple Tx with just Conway DRep Certs. 
Added RootTarget for DRepPulser. Introduced idea of RootTarget for Pairs. 
Added Virtual RootTarget like Lensed with additional name. 
Added a new HasTrace instances, simplified MockChain 
Added STS (MOCKCHAIN era) instance and helper functions. 
Fixed pot Equal obligation in Pool Reap and problems with tracking obligations. 
Made 'There are no "stakepools to choose an issuer from" a discard. 
Added property tests on traces with simple Tx with certs on DReps.

Fixes #3610 

# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
```
Comment (Soupstraw) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs]:
```suggestion
```

This looks redundant

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs [hunk 2]
```diff
@@ -183,3 +202,29 @@ strictSeqListL = lens toList (\_ y -> fromList y)
...
+pairL xy ab = lens getter setter
+  where
+    getter (x, a) = (x ^. xy, a ^. ab)
+    setter (x, a) (y, b) = (x & xy .~ y, a & ab .~ b)
+
+firstL :: Lens' (a, b) a
+firstL = lens fst (\(_, y) a -> (a, y))
+
+secondL :: Lens' (a, b) b
+secondL = lens snd (\(x, _) b -> (x, b))
```
Comment (Soupstraw) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: There are already lenses called `_1` and `_2` that do the same thing

Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: I removed them

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs (hunk 1 above)
Comment (TimSheard) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs]: It was!  good catch

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  , Rng goDelegs `Subset` poolHashUniv
+  , Sized (Range 1 2) goPools
+  , Dom goPools `Subset` poolHashUniv
+  , Random snapShotFee
+  , SumsTo (Right (1 % 1000)) (Lit RationalR 1) EQL [ProjMap RationalR individualPoolStakeL markPoolDistr]
+  , Sized (Range 2 6) markPoolDistr
+  ]
+
+newepochstatePreds :: Era era => Proof era -> [Pred era]
+newepochstatePreds _proof =
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs]:
```suggestion
newEpochStatePreds :: Era era => Proof era -> [Pred era]
newEpochStatePreds _proof =
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs [hunk 4]
```diff
@@ -111,12 +145,12 @@ pstateGenPreds _ =
...
-      -- But since it is so intimately tied to regPools we define it here
-      -- Alternately we could put this in NewEpochState, and insist that pStateStage
-      -- preceed newEpochStateStage
-      -- , Dom regPools :=: Dom poolDistr
-      -- , SumsTo (Right (1 % 1000)) (Lit RationalR 1) EQL [ProjMap RationalR individualPoolStakeL poolDistr]
+  , -- poolDistr not needed in PState, but is needed in NewEpochState
+    -- But since it is so intimately tied to regPools we define it here
+    -- Alternately we could put this in NewEpochState, and insist that pStateStage
+    -- preceed newEpochStateStage
+    Dom regPools :=: Dom poolDistr
```
Comment (MaximilianAlgehed) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: I would be very happy if we could move these constraints to where they are needed instead. Having variables used in constraints that don't relate to that type can cause some real headaches in properties.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 5]
```diff
@@ -390,7 +391,9 @@ prop_GOVCERT sub =
   stsProperty @"GOVCERT"
     sub
     (\PropEnv {..} -> ConwayGovCertEnv pePParams peEpochNo)
-    (\_ -> genShrinkFromConstraints conwayProof sub vstatePreds (const []) vstateT)
+    (\_ -> genShrinkFromConstraints conwayProof sub (\p -> Random (Vars.drepDeposit p) : vstatePreds p) (const []) vstateT)
```
Comment (MaximilianAlgehed) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs:394]: Does this fix my comment above?

Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs:394]: No wait derp! Sorry, my bad!

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs [hunk 6]
```diff
@@ -165,6 +177,13 @@ delegationsUMapL = lens sPoolMap delta
       where
         split = splitUMap um
 
+drepUMapL :: Lens' (UMap c) (Map (Credential 'Staking c) (DRep c))
+drepUMapL = lens dRepMap delta
+  where
+    delta um new = unSplitUMap (split {spDRep = new})
+      where
+        split = splitUMap um
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: Why extra binding?
```suggestion
    delta um new = unSplitUMap ((splitUMap um) {spDRep = new})
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs [hunk 7]
```diff
@@ -45,53 +47,56 @@ enactStateGenPreds p =
...
   , Random prevHardFork
   , Random prevConstitution
   , Random prevCommittee
-  , Random prevProposals
-  , Random prevDRepDistr
-  , Random currProposals
-  , Random ratifyState
-  , MetaSize (SzRng 90 (usNumPreUtxo usize)) utxoSize -- must be bigger than sum of (maxsize inputs 10) and (mazsize collateral 3)
+  , Sized (Range 0 1) currProposals
+  , proposalDeposits :<-: (Constr "sumActionStateDeposits" (\x -> List.foldl' (<>) (Coin 0) (fmap gasDeposit x)) :$ (Simple currProposals))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs]:
```suggestion
  , proposalDeposits :<-: (Constr "sumActionStateDeposits" (foldMap gasDeposit x) :$ Simple currProposals)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+import Test.Cardano.Ledger.Constrained.Vars
+import Test.Cardano.Ledger.Generic.PrettyCore (pcEpochState, pcNewEpochState)
+import Test.Cardano.Ledger.Generic.Proof
+import Test.QuickCheck
+import Test.Tasty (TestTree, defaultMain)
+
+-- ===========================================================
+
+epochstatePreds :: EraGov era => Proof era -> [Pred era]
+epochstatePreds _proof =
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs:37]:
```suggestion
epochStatePreds :: EraGov era => Proof era -> [Pred era]
epochStatePreds _proof =
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs [hunk 9]
```diff
@@ -51,6 +52,16 @@ nonNegativeInterval r = case (boundRational @NonNegativeInterval r) of
...
 
+{- Preds we want to be True in pparams
+  [ Sized (AtLeast 1) (maxBHSize proof)
+  , Sized (AtLeast 1) (maxTxSize proof)
+  , SumsTo (Right 1) (maxBBSize proof) LTE [One (maxBHSize proof), One (maxTxSize proof)]
+  , Component
+      (Right (pparams proof))
+      [field pp (maxTxSize proof), field pp (maxBHSize proof), field pp (maxBBSize proof)]
+  ]
+-}
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs]: Dead code?
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs [hunk 10]
```diff
@@ -106,9 +120,16 @@ pParamsPreds p =
...
-          PParamsBabbageToConway ->
+          PParamsBabbageToBabbage ->
+            [ extract (maxTxExUnits p) (pparams p)
+            , extract (collateralPercentage p) (pparams p)
+            ]
+          PParamsConwayToConway ->
             [ extract (maxTxExUnits p) (pparams p)
             , extract (collateralPercentage p) (pparams p)
+            , extract (drepDeposit p) (pparams p)
+            , extract (drepDeposit p) (pparams p)
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs]: It doesn't look like it was intended to be listed twice, no?
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs [hunk 11]
```diff
@@ -779,6 +784,37 @@ genTxAndLedger sizes proof = do
...
+        >>= ledgerStateStage sizes proof
+        >>= txBodyStage sizes proof
+        >>= epochStateStage proof
+        >>= newEpochStateStage proof
+      )
+  env0 <- monadTyped $ substToEnv subst emptyEnv
+  env1 <- monadTyped $ adjustFeeInput env0
+  env2 <- monadTyped $ adjustColInput env1
+  newepochst <- monadTyped $ runTarget env2 (newEpochStateT proof)
+  (TxF _ tx) <- monadTyped (findVar (unVar txterm) env2)
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: :slightly_smiling_face: 
```suggestion
  TxF _ tx <- monadTyped (findVar (unVar txterm) env2)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs [hunk 12]
```diff
@@ -779,6 +784,37 @@ genTxAndLedger sizes proof = do
...
+        >>= universeStage sizes proof
+        >>= vstateStage proof
+        >>= pstateStage proof
+        >>= dstateStage proof
+        >>= certsStage proof
+        >>= ledgerStateStage sizes proof
+        >>= txBodyStage sizes proof
+        >>= epochStateStage proof
+        >>= newEpochStateStage proof
+      )
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs:801]:
```suggestion
     pure emptySubst
        >>= pParamsStage proof
        >>= universeStage sizes proof
        >>= vstateStage proof
        >>= pstateStage proof
        >>= dstateStage proof
        >>= certsStage proof
        >>= ledgerStateStage sizes proof
        >>= txBodyStage sizes proof
        >>= epochStateStage proof
        >>= newEpochStateStage proof
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs [hunk 13]
```diff
@@ -779,6 +784,37 @@ genTxAndLedger sizes proof = do
...
+  env2 <- monadTyped $ adjustColInput env1
+  newepochst <- monadTyped $ runTarget env2 (newEpochStateT proof)
+  (TxF _ tx) <- monadTyped (findVar (unVar txterm) env2)
+  pure (newepochst, tx, env2)
+
+demoTxNes :: IO ()
+demoTxNes = do
+  let proof = Conway Standard
+  (nes, _tx, env) <- generate $ genTxAndNewEpoch def proof
+  putStrLn (show (psNewEpochState proof nes))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs:813]:
```suggestion
  print (psNewEpochState proof nes)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs [hunk 14]
```diff
@@ -130,11 +130,14 @@ demo mode = do
           >>= (\subst -> monadTyped $ substToEnv subst emptyEnv)
       )
   lstate <- monadTyped $ runTarget env (ledgerStateT proof)
+  let _env2 = getTarget lstate (ledgerStateT proof) emptyEnv
   when (mode == Interactive) $ putStrLn (show (pcLedgerState proof lstate))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs:137]:
```suggestion
  when (mode == Interactive) $ print (pcLedgerState proof lstate)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+          >>= vstateStage proof
+          >>= pstateStage proof
+          >>= dstateStage proof
+          >>= ledgerStateStage def proof
+          >>= epochStateStage proof
+          >>= (\subst -> monadTyped $ substToEnv subst emptyEnv)
+      )
+  epochstate <- monadTyped $ runTarget env (epochStateT proof)
+  let env2 = getTarget epochstate (epochStateT proof) emptyEnv
+  when (mode == Interactive) $ putStrLn (show (pcEpochState proof epochstate))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs]:
```suggestion
  when (mode == Interactive) $ print (pcEpochState proof epochstate)
```

Seems like a pattern that is used a lot, maybe adding a helper function for it?
```haskell
printInteractive mode = when (mode == Interactive) . print
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+          >>= pstateStage proof
+          >>= dstateStage proof
+          >>= ledgerStateStage def proof
+          >>= epochStateStage proof
+          >>= newEpochStateStage proof
+          >>= (\subst -> monadTyped $ substToEnv subst emptyEnv)
+      )
+  newepochstate <- monadTyped $ runTarget env (newEpochStateT proof)
+  let env2 = getTarget newepochstate (newEpochStateT proof) emptyEnv
+  when (mode == Interactive) $ putStrLn (show (pcNewEpochState proof newepochstate))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs:126]:
```suggestion
  when (mode == Interactive) $ print (pcNewEpochState proof newepochstate)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs [hunk 17]
```diff
@@ -286,12 +330,19 @@ mainD seed = defaultMain $ testIO "Testing DState Stage" (demoD Interactive seed
 
 -- ===============================================
 
+tell :: String -> (Subst era -> a) -> (Subst era -> a)
+tell s f (Subst m) =
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: `tell` is a name of a very popular function in `transformers` library. I'd suggest to rename it in order to avoid confusion.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs [hunk 18]
```diff
@@ -286,12 +330,19 @@ mainD seed = defaultMain $ testIO "Testing DState Stage" (demoD Interactive seed
...
+  case Map.lookup "drepState" m of
+    Nothing -> trace (s ++ " NO") (f (Subst m))
+    Just x -> trace (s ++ " YES " ++ show x) (f (Subst m))
+
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: Let's snap a warning on it to prevent it actually being used somewhere in future PRs
```suggestion
{-# WARNING tell "This function suppose to be used for debugging only" #-}
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Trace/DrepCertTx.hs [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+import Cardano.Ledger.Hashes (EraIndependentTxBody)
+
+-- =========================================================================
+
+-- | Fix the first Outputs field in a [TxBodyField], by applying the delta Coin to the first Output in that Outputs field
+--   This is used to compensate for certificates which make deposits, and this Coin must come from somewhere, so it is
+--   added (subtracted if the Coin is negative) from the first TxOut.
+fixOutput :: EraTxOut era => Coin -> [TxBodyField era] -> [TxBodyField era]
+fixOutput _ [] = []
+fixOutput delta (Outputs' (txout : more) : others) = (Outputs' (updateWithLens coinTxOutL (<+> delta) txout : more)) : others
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Trace/DrepCertTx.hs]: Line is too long, tons of redundant parenthesis and instead of `updateWithLens` use either `%~ f` or in this case [`<>~`](https://hackage.haskell.org/package/microlens-0.4.13.1/docs/Lens-Micro.html#v:-60--62--126-) is even better
```suggestion
fixOutput delta (Outputs' (txout : more) : others) = 
  Outputs' (txout & coinTxOutL <>~ delta) : more : others
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs [hunk 20]
```diff
@@ -47,6 +53,9 @@ mmL = lens mm (\ds u -> ds { mm = u })
 
 -}
 
+updateWithLens :: Lens' a b -> (b -> b) -> a -> a
+updateWithLens l f a = a & l .~ f (a ^. l)
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: There is already an operator for this: [`%~`](https://hackage.haskell.org/package/microlens-0.4.13.1/docs/Lens-Micro.html#v:-37--126-)
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs [hunk 21]
```diff
@@ -370,6 +388,7 @@ showT (f :$ x) = "(" ++ showT f ++ " " ++ showL pp " " (args x) ++ ")"
 showT (Invert nm _ _) = nm
 showT (Shift x _) = "(Shift " ++ showT x ++ ")"
 showT (Mask x) = showT x
+showT (Virtual _ y _) = show y -- "(Virtual " ++ show x ++ " " ++ y ++ ")"
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs]: Dead code?
```suggestion
showT (Virtual _ y _) = show y
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs [hunk 22]
```diff
@@ -753,13 +757,17 @@ dispatch v1 preds | Just (If tar x y) <- List.find isIf preds = do
   b <- simplifyTarget tar
   let others = filter (not . isIf) preds
   if b then dispatch v1 (x : others) else dispatch v1 (y : others)
-dispatch v1@(V nam r1 _) preds = explain ("Solving for variable " ++ nam ++ show preds) $ case preds of
+dispatch v1@(V nam r1 _) preds = explain ("Solving for variable " ++ nam ++ "\n" ++ unlines (map show preds)) $ case preds of
   [Var v2 :=: Lit r2 t] | Name v1 == Name v2 -> do
     Refl <- sameRep r1 r2
     pure (pure t)
-  [Lit r2 t :=: Var v2] | Name v1 == Name v2 -> do
+  [Lit r2 t :=: Var v2] | (Name v1) == Name v2 -> do
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs]: Unnecessary change
```suggestion
  [Lit r2 t :=: Var v2] | Name v1 == Name v2 -> do
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs [hunk 23]
```diff
@@ -753,13 +757,17 @@ dispatch v1 preds | Just (If tar x y) <- List.find isIf preds = do
...
     Refl <- sameRep r1 r2
     pure (pure t)
+  [expr :=: Var v2] | (Name v1) == Name v2 -> do
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs]: Redundant parenthesis
```suggestion
  [expr :=: Var v2] | Name v1 == Name v2 -> do
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs [hunk 24]
```diff
@@ -183,3 +202,23 @@ strictSeqListL = lens toList (\_ y -> fromList y)
 
 mapCompactFormCoinL :: Lens' (Map a (CompactForm Coin)) (Map a Coin)
 mapCompactFormCoinL = lens (Map.map fromCompact) (\_ y -> Map.map compactCoinOrError y)
+
+drepDelegationUMapL :: Lens' (UMap c) (Map (Credential 'Staking c) (DRep c))
+drepDelegationUMapL = lens dRepMap delta
+  where
+    delta um new = unSplitUMap (split {spDRep = new})
+      where
+        split = splitUMap um
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: Looks identical to `drepUMapL` above ^
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs [hunk 25]
```diff
@@ -130,11 +130,14 @@ demo mode = do
           >>= (\subst -> monadTyped $ substToEnv subst emptyEnv)
       )
   lstate <- monadTyped $ runTarget env (ledgerStateT proof)
+  let _env2 = getTarget lstate (ledgerStateT proof) emptyEnv
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs]: Dead code?
```suggestion
```

**Review by lehins — APPROVED** [review]
LGTM :+1:

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs (hunk 3 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs (hunk 4 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: So the best I could do was move them to the Ledger stage, as it needs poolDistr to construct a valid pulser.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs (hunk 6 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs (hunk 7 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs (hunk 8 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs:37]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs (hunk 9 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs]: This comment is meant as a statement of intent
 The actual code achieves this algorithmically, but we want this declaration of what should be true

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs (hunk 21 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs (hunk 22 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs (hunk 23 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs (hunk 24 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: Yeah, not sure how that happened. There is only one version now.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs (hunk 25 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs]: Actually we are testing env2, but while tracking down some bug, I switched to the earlier 'env', it is back where it should be. Good catch.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs (hunk 10 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs (hunk 11 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs (hunk 12 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs:801]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs (hunk 15 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs (hunk 17 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: Actually that was a debugging function that is no longer needed.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs (hunk 18 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: I removed it

**Review by MaximilianAlgehed — APPROVED** [review]

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs (hunk 20 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Lenses.hs]: deleted

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Trace/DrepCertTx.hs (hunk 19 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Trace/DrepCertTx.hs]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs (hunk 16 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs:126]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs (hunk 14 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs:137]: done

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs (hunk 13 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs:813]: done
